XiEditor/Document.swift

```diff
@@ -150,10 +150,14 @@ class Document: NSDocument {
         //TODO: save operations should report success, and we should pass any errors to the completion handler
         completionHandler(nil)
     }
-    
+
+    // Document.close() can be called multiple times (on window close and application terminate)
     override func close() {
-        super.close()
-        Events.CloseView(viewIdentifier: coreViewIdentifier!).dispatch(dispatcher!)
+        if let identifier = self.coreViewIdentifier {
+            self.coreViewIdentifier = nil
+            Events.CloseView(viewIdentifier: identifier).dispatch(dispatcher!)
+            super.close()
```

Comment:
this is a very good question, to which I don't have an entirely satisfactory answer. The apple documentation does not mention calling super in your implementation *at all*, although super is called in some sample code. I think there's a reasonable argument for moving it outside. Given how little we are actually involved in the NSDocument lifecycle stuff (i.e. we handle all opening/closing etc in core) I'm not inclined to worry here? But moving it outside might be more correct.